File: src/org/dyndns/fzoli/rccar/model/bridge/StorageList.java
package org.dyndns.fzoli.rccar.model.bridge;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import org.dyndns.fzoli.rccar.bridge.config.Permissions;
import org.dyndns.fzoli.rccar.model.controller.HostList;
import org.dyndns.fzoli.rccar.model.host.HostData;
import org.dyndns.fzoli.socket.process.impl.MessageProcess;

/**
 *
 * Az online járművek és vezérlők adatait tartalmazó lista.
 * @author zoli
 */
public class StorageList {
    
    /**
     * A jármű-adattárolók szálbiztos konténere.
     */
    private static final List<HostStorage> HOSTS = Collections.synchronizedList(new ArrayList<HostStorage>());
    
    /**
     * A vezérlő-adattárolók szálbiztos konténere.
     */
    private static final List<ControllerStorage> CONTROLLERS = Collections.synchronizedList(new ArrayList<ControllerStorage>());

    /**
     * A vezérlők tárolói.
     * Használata iteráláshoz pl. üzenet küldése több vezérlőnek
     */
    public static List<ControllerStorage> getControllerStorageList() {
        return CONTROLLERS;
    }
    
    /**
     * A járművek tárolói.
     * Használata iteráláshoz pl. üzenet küldése több járműnek
     */
    public static List<HostStorage> getHostStorageList() {
        return HOSTS;
    }
    
    /**
     * A jármű azonosítója alapján megkeresi a hozzá tartozó tárolót.
     * @return null, ha nem létezik a tároló
     */
    public static HostStorage findHostStorageByName(String name) {
        return findStorageByName(name, HOSTS);
    }
    
    /**
     * A vezérlő azonosítója alapján megkeresi a hozzá tartozó tárolót.
     * @return null, ha nem létezik a tároló
     */
    public static ControllerStorage findControllerStorageByName(String name) {
        return findStorageByName(name, CONTROLLERS);
    }
    
    /**
     * Létrehoz és tárol egy {@link ControllerStorage} objektumot, vagy beállítja a már létezőt és azzal tér vissza.
     * @param messageProcess a vezérlő jelenlegi üzenetküldője
     */
    public static ControllerStorage createControllerStorage(MessageProcess messageProcess) {
        ControllerStorage s = findControllerStorageByName(messageProcess.getLocalCommonName());
        if (s == null) {
            s = new ControllerStorage(messageProcess);
            getControllerStorageList().add(s);
        }
        else {
            s.setMessageProcess(messageProcess);
        }
        return s;
    }
    
    /**
     * Létrehoz és tárol egy {@link HostStorage} objektumot, vagy beállítja a már létezőt és azzal tér vissza.
     * @param messageProcess a host jelenlegi üzenetküldője
     * @param data a beállított adatmodel
     */
    public static HostStorage createHostStorage(MessageProcess messageProcess, HostData data) {
        HostStorage s = findHostStorageByName(messageProcess.getLocalCommonName());
        if (s == null) {
            s = new HostStorage(messageProcess);
            getHostStorageList().add(s);
        }
        else {
            s.setMessageProcess(messageProcess);
        }
        s.getHostData().update(data);
        return s;
    }
    
    /**
     * Az online járművek listáját generálja le.
     */
    public static HostList createHostList(String controllerName) {
        HostList l = new HostList();
        List<String> ls = l.getHosts();
        for (HostStorage s : getHostStorageList()) {
            if (HostStorage.isHostConnected(s) && Permissions.getConfig().isEnabled(s.getName(), controllerName)) {
                ls.add(s.getName());
            }
        }
        return l;
    }
    
    /**
     * Név alapján megkeresi a tárolót a megadott listában.
     * @param name a név paraméter
     * @param list a lista, amiben keresni kell
     */
    private static <S extends Storage> S findStorageByName(String name, List<S> list) {
        if (name != null) {
            for (S s : list) {
                if (name.equals(s.getName())) return s;
            }
        }
        return null;
    }
    
}

File: cpp/RCServer/StorageList.cpp
/* 
 * File:   StorageList.cpp
 * Author: zoli
 * 
 * Created on 2013. szeptember 20., 16:33
 */

#include "StorageList.h"
#include "HostStorage.h"
#include "ControllerStorage.h"

StorageList::HostStorageVector StorageList::hosts;
StorageList::ControllerStorageVector StorageList::controllers;

StorageList::HostStorageVector& StorageList::getHostStorages() {
    return hosts;
}

StorageList::ControllerStorageVector& StorageList::getControllerStorages() {
    return controllers;
}

StorageList::HostStorageType* StorageList::findHostStorageByName(std::string name) {
    for (HostStorageVector::iterator it = hosts.begin(); it != hosts.end(); it++) {
        HostStorageType* s = *it;
        if (name == s->getName()) return s;
    }
    return NULL;
}

StorageList::ControllerStorageType* StorageList::findControllerStorageByName(std::string name) {
    for (ControllerStorageVector::iterator it = controllers.begin(); it != controllers.end(); it++) {
        ControllerStorageType* s = *it;
        if (name == s->getName()) return s;
    }
    return NULL;
}

void StorageList::freezeControllerStorage(SSLProcess* p) {
    ControllerStorageType* s = findControllerStorageByName(p->getSocket()->getClientName());
    if (s) s->setMessageProcess(NULL);
}

void StorageList::freezeHostStorage(SSLProcess* p) {
    HostStorageType* s = findHostStorageByName(p->getSocket()->getClientName());
    if (s) s->setMessageProcess(NULL);
}

StorageList::ControllerStorageType* StorageList::createControllerStorage(MessageProcess* p) {
    ControllerStorageType* s = findControllerStorageByName(p->getSocket()->getClientName());
    if (s == NULL) {
        s = new ControllerStorage(p);
        controllers.push_back(s);
    }
    else {
        s->setMessageProcess(p);
    }
    return s;
}

StorageList::HostStorageType* StorageList::createHostStorage(MessageProcess* p, HostData* d) {
    HostStorageType* s = findHostStorageByName(p->getSocket()->getClientName());
    if (s == NULL) {
        s = new HostStorage(p);
        hosts.push_back(s);
    }
    else {
        s->setMessageProcess(p);
    }
    HostStorage* hs = (HostStorage*) s;
    hs->getHostData().update(d);
    return s;
}

HostList StorageList::createHostList(std::string controllerName) {
    HostList l;
    for (HostStorageVector::iterator it = hosts.begin(); it != hosts.end(); it++) {
        HostStorage* s = (HostStorage*) *it;
        if (s->isConnected()) {
            l.addHost(s->getName());
        }
    }
    return l;
}
